### DISCUSSION TITLE
Shared mutable object in crossroads property handler

### DISCUSSION THREAD

**Discussion by Taha-Firoz** [diwic/dbus-rs#348] [Q&A] [answered]
I need to have a shared mutable reference to the underlying object in the cross roads handler. In order to achieve that I've wrapped the object in an `Arc<Mutex<...>>` but when I want to access the object inside of a property handler I can't seem to get it working. I don't know if the Arc is exactly necessary. 

I'm using `tokio::sync::Mutex` which requires an await to lock, 
```
iface_builder.property("Device")
            .get(|prop, device:&mut Arc<Mutex<Manager>>| { 
                  ...  
                }
            );
```
This is fine but when I actually try to lock the mutex to use it I need to await it. So I saw a method called `get_async` but I keep getting this error when I try to do it
```
iface_builder.property("Device")
            .get_async(|prop, device:&mut Arc<Mutex<Manager>>| { 
                    async move{
                        let s = device.lock().await;
                        s.is_online
                    }
            );
```
I get this error
```
type mismatch resolving `<impl futures::Future as futures::Future>::Output == PhantomData<(_, ())>`
expected enum `Option<Vec<std::string::String>>`
 found struct `PhantomData<(_, ())>`
```
I don't really understand what to do here now :/ any help would be appreciated. I'm trying to wrap a DBus interface over a service that's running a separate tokio task which is why I think I need the `Arc<Mutex<...>>`

**Comment by surban** [comment, marked as answer]
You need to call `PropContext::reply` instead of returning the value directly, i.e. `prop.reply(Ok(s.is_online))` in the last line of your async block.

**Reply by surban** [reply]
You can check out a similar implementation for async D-Bus methods at https://github.com/bluez/bluer/blob/8f69f851ea25b0748099f2d72cc263839473728a/bluer/src/gatt/local.rs#L761

**Reply by Taha-Firoz** [reply]
@surban The error is a little different but it's still there
```
iface_builder.property("Device")
            .get_async(|prop, device| async move{ 
                        let s = device.lock().await;
                        return prop.reply(Ok(s.is_online.unwrap()))
                }
            );
```
```
type mismatch resolving `<impl futures::Future as futures::Future>::Output == PhantomData<(_, ())>`
expected struct `PhantomData<Vec<std::string::String>>`
   found struct `PhantomData<(_, ())>`
```

**Reply by diwic** [reply]
Hm, this does indeed seem to be a bug. I just pushed a commit that should make it work (with `reply`) as @surban correctly suggests. Could you try it?

**Reply by Taha-Firoz** [reply]
@diwic I don't understand how to specify the dependency for crossroads to git because the structure of the repository.

**Reply by Taha-Firoz** [reply]
Doing this `dbus-crossroads = { git = "https://github.com/diwic/dbus-rs" }` which should resolve the path automatically, gives me this 

```
   Updating git repository `https://github.com/diwic/dbus-rs`
error: failed to select a version for `libdbus-sys`.
    ... required by package `dbus v0.9.3 (https://github.com/diwic/dbus-rs#06f8edab)`
    ... which is depended on by `dbus-crossroads v0.4.0 (https://github.com/diwic/dbus-rs#06f8edab)`
    ... which is depended on by `dbus-test v0.1.0 (.../dbus-test)`
versions that meet the requirements `^0.2` are: 0.2.1

the package `libdbus-sys` links to the native library `dbus`, but it conflicts with a previous package which links to `dbus` as well:
package `libdbus-sys v0.2.1`
    ... which is depended on by `dbus v0.9.3`
    ... which is depended on by `dbus-test v0.1.0 (.../dbus-test)`

failed to select a version for `libdbus-sys` which could resolve this conflict
```

**Reply by diwic** [reply]
@Taha-Firoz I'm not sure either.

Two things to try:
1) If you make `dbus-crossroads` point to git, maybe `dbus` (if you have that as a dependency) should also point to git. Same goes for `dbus-tokio` etc if you have other crates from the repo.
2) You can clone the repo and then do `dbus-crossroads = { path = "/somewhere/dbus-rs/dbus-crossroads" }`.

Or both strategies combined.

**Reply by Taha-Firoz** [reply]
@diwic  Okay that did work and now that error's gone, however this new error's come up.

```
error[E0495]: cannot infer an appropriate lifetime due to conflicting requirements
  --> src/main.rs:46:32
   |
46 |                       async move {
   |  ________________________________^
47 | |                         let s = device.read().await;
48 | |     
49 | |                         prop.reply(Ok(s.is_online))
50 | |                     }
   | |_____________________^
   |
note: first, the lifetime cannot outlive the anonymous lifetime #1 defined on the body at 45:28...
  --> src/main.rs:45:28
   |
45 |                   .get_async(|mut prop, device| 
   |  ____________________________^
46 | |                     async move {
47 | |                         let s = device.read().await;
48 | |     
49 | |                         prop.reply(Ok(s.is_online))
50 | |                     }
   | |_____________________^
note: ...so that the types are compatible
  --> src/main.rs:46:32
   |
46 |                       async move {
   |  ________________________________^
47 | |                         let s = device.read().await;
48 | |     
49 | |                         prop.reply(Ok(s.is_online))
50 | |                     }
   | |_____________________^
   = note: expected `(&mut tokio::sync::RwLock<Manager>, PropContext)`
              found `(&mut tokio::sync::RwLock<Manager>, PropContext)`
   = note: but, the lifetime must be valid for the static lifetime...
note: ...so that the type `impl futures::Future` will meet its required lifetime bounds
  --> src/main.rs:45:18
   |
45 |                 .get_async(|mut prop, device| 
```

For reference this is my handler, I changed the device from being in an `Arc<Mutex<...>` to a `RwLock<...>` could that be causing the problem
```
iface_builder
                .property("Device")
                .get_async(|mut prop, device| 
                    async move {
                        let s = device.read().await;
    
                        prop.reply(Ok(s.is_online))
                    }
                );
```
The only reason I continue this thread is this example isn't that much different from whats going on the `method_cr_async` where this would probably work in. The example is super minimal and is just a struct wrapped in an `RwLock` with these handlers.

**Reply by diwic** [reply]
Okay, so this is more of a "async closures in Rust" problem and less of a "dbus-crossroads" problem. But long story short, you probably want your device in an `Arc` and do something like:

    .get_async(|mut prop, device|
        let d = device.clone();
        async move {
            let s = d.read().await;
            prop.reply(Ok(s.is_online))
        }
    );

**Reply by diwic** [reply]
...and this is because `device` is only mutably borrowed until the first await point. It can only be that, because when awaiting, get_async can be called in parallel which means that the second call to get_async will also need its mutable borrow of `device`.

**Reply by Taha-Firoz** [reply]
Okay everything's compiling now!

**Reply by Taha-Firoz** [reply]
@diwic till when do you think you'll publish the crate with the fix? I really appreciate all the help though, thank you!

**Reply by diwic** [reply]
How about ... *drum roll* ... right now? Enjoy!
